pyhf/optimize/__init__.py

```diff
@@ -1,24 +1,39 @@
-from .opt_scipy import scipy_optimizer
+class _OptimizerRetriever(object):
+    def __getattr__(self, name):
+        if name == 'scipy_optimizer':
+            from .opt_scipy import scipy_optimizer
 
-assert scipy_optimizer
+            assert scipy_optimizer
+            self.scipy_optimizer = scipy_optimizer
```

Comment:
what's the assignment for?

Reply:
When assigned, the custom `__getattr__` won't be called if the attribute already exists. The `__getattr__()` is only called when the attribute doesn't exist. This is the lazy-loading part... so once it's loaded, we just add it in and we're done.

Reply:
shouldn't the import be lazy anyways?

```
import foo
import foo
```

should not really import twice, no?

Reply:
I think you're getting confused. In any other piece of code, we do something like `import pyhf.tensor`. When this is called once, all it does is load the object defined in that module. It doesn't have any backends loaded. When `pyhf.tensor.numpy_backend` is called, it looks through the existing attributes of the object, sees that it doesn't exist, then tries to import it. When we can import it, we then set that attribute on the object to bypass the `__getattr__` check which only occurs if the object didn't have the attribute in the first place.

Reply:
I guess I am, I'd think that 

```
def lazyload():
    from bla import X
    return X

#nothing loaded until now

X = lazyload()
X = lazyload()
```
would achieve the same thing, without assignment

Reply:
This isn't about the import loading. The `__getattr__` is only run if it cannot find the attribute on the tensor object we're making. Since the object should just represent what we import, we should set the attribute on it so that when python goes through all attributes to match the call, it short-circuits instead of going to our `__getattr__` which should be fast since it doesn't actually import again.

We're not trying to speed up the import here, but the `getattr` check.

Reply:
ok but clearly the bottleneck is in the import and not in the getattr check no? Maybe I'm overlooking something, how much are we saving?

Reply:
I've added comments to clarify.